### PR TITLE
It is no longer possible to target TokenHolder itself

### LINKED ISSUE #76: TokenHolder: Bypassing Multisig with executeRule should not be possible
The `TokenHolder` contract serves mainly as a multisig wallet to give users a strong wallet to manage their tokens. Users are not forced to make use of the multisig features, however some users might choose to do so. While analyzing the different parts of the `TokenHolder` contract, it was found that the multisig requirements can be bypassed by anybody who can sign a transaction with an authorized `ephemeralKey`. This is due to how the multisig wallet will execute confirmed transactions and how the rule execution is implemented.

As can be seen in the two code snippets below, the multisig `executeTransaction` function uses `call()` to execute the internal multisig functions, while `executeRule` also uses `call()` but with a different intention. To execute rules an `ephemeralKey` has to be authorized by the multisig wallet, however depending on the setup a user or even the organizer might be able to sign data with that `ephemeralKey`. Essentially the control over the `ephemeralKey` is equal to the control of the required wallets for the `multisig` features. This means an attacker could sign a `executeRule` call with the `TokenHolder` contract as `_to` and the `_data` to call one of the internal multisig functions such as `addWallet` or `changeRequirement`.

*Affected Files:*
```solidity
openst-contracts/contracts/TokenHolder.sol
openst-contracts/contracts/MultiSigWallet.sol
```

*Affected Code:*
```solidity
function executeRule(address _to,bytes _data,uint256 _nonce,uint8 _v,bytes32 _r, bytes32 _s) public payable returns (bool executionStatus_)
{
    bytes32 messageHash = bytes32(0);
    address ephemeralKey = address(0);
    (messageHash, ephemeralKey) = verifyExecutableTransaction(
EXECUTE_RULE_CALLPREFIX, _to, _data, _nonce, _v, _r, _s );
    // [...]
    // solium-disable-next-line security/no-call-value
    executionStatus_ = _to.call.value(msg.value)(_data);
    // [...]
}
```

*Affected Code:*
```solidity
function executeTransaction(uint256 _transactionID) public onlyWallet
    transactionExists(_transactionID)
    transactionIsConfirmedBy(_transactionID, msg.sender)
    transactionIsNotExecuted(_transactionID)
{
    if (isTransactionConfirmed(_transactionID)) {
        Transaction storage t = transactions[_transactionID];
        // solium-disable-next-line security/no-low-level-calls
        if (t.destination.call(t.data)) {
            t.executed = true;
            emit TransactionExecutionSucceeded(_transactionID);
        } else {
            t.executed = false;
            emit TransactionExecutionFailed(_transactionID);
        }
    }
}
```

*PoC:*
The following proof of concept code signs the call to the internal multisig function `changeRequirement(1)`. The from and to address is set to the `TokenHolder` contract. The resulting signed data can be used with `executeRule` to change the multisig wallet requirement, without having to go through the multisig confirmation steps.

```js
console.log('** Sign the data as per EIP 1077.');
let eip1077SignedData = ephemeralKeyAccount.signEIP1077Transaction({
  from: '0x9240ddc345d7084cc775eb65f91f7194dbbb48d8',
  to: '0x9240ddc345d7084cc775eb65f91f7194dbbb48d8',
  value: 0,
  gasPrice: 0,
  gas: 0,
  data: '0xba51a6df0000000000000000000000000000000000000000000000000000000000000001',
  nonce: 1,
  callPrefix: '0x59793b00'
});
```

As can be seen from the `changeRequirement` example above, this could be used to add, remove or replace wallets in the list, thus fully takeover the `TokenHolder` multisig wallet.
A simple fix would simply prevent `executeRule` from calling its own contract, which can be achieved by adding a check in `executeRule` that `_to` must not be `address(this)`.

### PR DESCRIPTION
**Pull request by schemar**
An attacker can no longer sign an executeRule call with the TokenHolder
contract as `_to` and the `_data` to call one of the internal multisig
functions such as for examyle ` addWallet` or `changeRequirement`.

See #76 for details.

Fixes #76

### CODE REVIEW

File: contracts/TokenHolder.sol [hunk 1]
```diff
@@ -361,6 +361,11 @@ contract TokenHolder is MultiSigWallet {
         private
         returns (bytes32 messageHash_, address key_)
     {
+        require(
+            _to != address(this),
+            "Target of a transaction cannot be TokenHolder itself."
+        );
+
```
Comment (abhayks1) [inline contracts/TokenHolder.sol]: The branch is not rebased with latest develop changes.
We should rebase with latest develop. Require for EIP20 token has been put at the start of executeRule function. I think we should put both requires together after rebasing with develop.

File: contracts/TokenHolder.sol [hunk 2]
```diff
@@ -361,6 +361,11 @@ contract TokenHolder is MultiSigWallet {
         private
         returns (bytes32 messageHash_, address key_)
     {
+        require(
+            _to != address(this),
+            "Target of a transaction cannot be TokenHolder itself."
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/TokenHolder.sol]: ✅

File: contracts/TokenHolder.sol (hunk 1 above)
Comment (schemar) [reply contracts/TokenHolder.sol]: ✅

**Comment by abhayks1** [comment]
Changes looks good after Whitelist Vs Blacklist discussion points. However there is travis build failure. 
cc: @schemar

**Comment by schemar** [comment]
I can confirm. But if you feel that someone else should review it as well, maybe just summon them? 😄

**Comment by abhayks1** [comment]
@jasonklein @pgev I have looked into this PR. Please look into it once and merge if it looks good to you also.

**Comment by jasonklein** [comment]
LGTM 🚀

**Pull request merged**